protocols/manage/broadcast.go

```diff
@@ -14,146 +16,82 @@ func init() {
-		dbg.Error(b.Info(), "failed to register channel:", err)
+	if b.tnIndex == -1 {
+		return nil, errors.New("Didn't find my TreeNode in the Tree")
 	}
-	if err := b.RegisterChannel(&b.okChan); err != nil {
-		dbg.Error(b.Info(), "failed to register channel:", err)
+	err := n.RegisterHandler(b.handleContactNodes)
+	if err != nil {
+		return nil, err
 	}
-
-	lists := b.Tree().List()
-	b.listNode = make(map[sda.TreeNodeID]*sda.TreeNode)
-	b.ackdNode = 0
-	b.done = make(chan bool, 1)
-	for _, tn := range lists {
-		if tn.Id.Equals(n.TreeNode().Id) {
-			continue
-		}
-		b.listNode[tn.Id] = tn
+	err = n.RegisterHandler(b.handleDone)
+	if err != nil {
+		return nil, err
 	}
-	go b.listen()
-	return b
-}
-
-// NewBroadcastRootProtocol is an abomination that should not exist - will
-// be killed with https://github.com/dedis/cothority/pull/325
-func NewBroadcastRootProtocol(n *sda.Node) (*Broadcast, error) {
-	b := new(Broadcast).init(n)
-	// it does not start yet.
 	return b, nil
 }
 
-func (b *Broadcast) listen() {
-	for {
-		select {
-		case msg := <-b.announceChan:
-			b.handleAnnounce(msg.TreeNode)
-		case msg := <-b.ackChan:
-			b.handleACK(msg.TreeNode)
-		case msg := <-b.okChan:
-			b.handleOk(msg.TreeNode)
-		case <-b.done:
-			return
-		}
-	}
-}
-
 // Start will contact everyone and makes the connections
 func (b *Broadcast) Start() error {
-	for _, tn := range b.listNode {
-		if err := b.SendTo(tn, &Announce{}); err != nil {
-			dbg.Error(b.Name(), "failed to send to", tn.Name(), err)
-		}
-	}
+	n := len(b.Tree().List())
+	b.repliesLeft = n * (n - 1) / 2
+	b.SendTo(b.Root(), &ContactNodes{})
 	dbg.Lvl3(b.Name(), "Sent Announce to everyone")
 	return nil
 }
 
 // handleAnnounce receive the announcement from another node
 // it reply with an ACK.
-func (b *Broadcast) handleAnnounce(tn *sda.TreeNode) {
-	if err := b.SendTo(tn, &ACK{}); err != nil {
-		dbg.Error(b.Info(), "failed to send ACK to", tn.Name(), err)
-	}
-}
-
-// It checks if we have sent an Announce to this treenode (hopefully yes^^)
-// if yes it checks if everyone has been ACK'd, if yes, it finishes.
-func (b *Broadcast) handleACK(tn *sda.TreeNode) {
-	if _, ok := b.listNode[tn.Id]; !ok {
-		dbg.Error(b.Name(), "Broadcast Received ACK from unknown treenode")
-	}
-
-	b.ackdNode++
-	if b.ackdNode == len(b.listNode) {
-		if !b.IsRoot() {
-			if err := b.SendTo(b.Tree().Root, &OK{}); err != nil {
-				dbg.Error(b.Info(), "failed to notify the root",
-					err)
+func (b *Broadcast) handleContactNodes(msg struct {
+	*sda.TreeNode
+	ContactNodes
+}) error {
+	dbg.Lvl3(b.Info(), "Received message from", msg.TreeNode.String())
+	if msg.TreeNode.Id == b.Root().Id {
+		dbg.Lvl3(b.Info(), "Contacting everybody")
+		// Connect to all nodes that are later in the TreeNodeList, but only if
+		// the message comes from root
+		for _, tn := range b.Tree().List()[b.tnIndex+1:] {
+			dbg.Lvl3("Connecting to", tn.String())
+			err := b.SendTo(tn, &ContactNodes{})
+			if err != nil {
+				return err
```

Comment:
In fact, having a handler does not necessarily have to return an error. `RegisterHandler` takes an interface, and we don't check for error, so a `dbg.Error` could do the trick.